akka-stream/src/main/scala/akka/stream/impl/io/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,125 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ ActorLogging, DeadLetterSuppression, Props }
+import akka.io.DirectByteBufferPool
+import akka.stream.actor.ActorPublisherMessage
+import akka.util.ByteString
+
+import scala.annotation.tailrec
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
```

Comment:
> 1) known chunk size – the read buffer, someone somewhere will set this value, it's not magic so people could end up depending on it even if not exposed as parameter – (does not concat multiple reads into

My point here that there must be a limit, either because you don't read ahead, and in this case you cannot return more than the read buffer contains, or because you read ahead, but you don't want to slurp the whole universe.

> 2) reads ahead, trying to lessen the delay from request to getting data. Maybe not needed - your points on when it would kick in and may be surprising are good ones

I believe in the read-ahead because it allows pipelining IO and CPU bound tasks. Even if we don't read ahead when having no demand, a request larger than 1 unit will imply a read-ahead anyway.

> 3) is an Actor. When I started this impl it was the JDK7 thingies (async) and stayed an actor. I can easily make it a Stage, however: in a stage I do not get the Request(N) explicitly (unless I missed something in StatefulStage API), and don't know how many chunks to read+collapse.

In the stage case you would not be able to batch, but the read-ahead will naturally arise from being a BoundaryStage at the end and having demands more than 1 from downstream. If it would be fused with other stages, then those stages would not get read-ahead, because they share the same synchronous island. I actually find this desirable, since when we will enable fusing support you can choose if you want pipelined async execution, or not pipelined sync execution.